### PR TITLE
TLSv13 official

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
For discussion, the RFC is published.  It's supported in JDK11 EA.  Conscrypt coming soon.

Biggest question is testing around inclusion in Modern or Restricted Connection Specs by default.  As a separate connection spec it likely causes churn because most (non FAANG) servers will fail for some time and fallback to 1.2.  But may cause some problems with 1.2 only servers that don't like the new handshake protocol (TLS version negotiation via attributes).

File: okhttp/src/main/java/okhttp3/ConnectionSpec.java [hunk 1]
```diff
@@ -83,11 +83,11 @@
...
 
   /**
-   * A TLS 1.3 only Connection Spec for testing. This will be eventually be exposed
+   * A TLS 1.3 only Connection Spec. This will be eventually be exposed
    * as part of MODERN_TLS or folded into the default OkHttp client once published and
-   * available in JDK11.
+   * available in JDK11 or Conscrypt.
    */
-  static final ConnectionSpec TLS_13 = new Builder(true)
+  public static final ConnectionSpec TLS_13 = new Builder(true)
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/ConnectionSpec.java]: Unsure about this in particular. Should we go more opinionated and just add TLS 1.3 to the modern set?

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/ConnectionSpec.java]: Probably yes.  I'll re-add with the ciphers, since they are now formalised.

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/ConnectionSpec.java]: There are 3 worrying results in the final set, particularly the google result.

```
Running tests using Jdk9Platform 11-ea+26
TLS1.3 only
https://enabled.tls13.com                java.net.ConnectException: Failed to connect to enabled.tls13.com/2400:cb00:2048:1:0:0:6810:7e22:443
https://www.howsmyssl.com/a/check        javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
https://tls13.cloudflare.com             java.net.ConnectException: Failed to connect to tls13.cloudflare.com/2400:cb00:2048:1:0:0:6813:c597:443
https://www.allizom.org/robots.txt       java.net.ConnectException: Failed to connect to www.allizom.org/2400:cb00:2048:1:0:0:6810:7e21:443
https://tls13.crypto.mozilla.org/        javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://tls.ctf.network/robots.txt       TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 404 169b
https://rustls.jbp.io/                   TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 200 416b
https://h2o.examp1e.net                  javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
https://mew.org/                         TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 137b
https://tls13.baishancloud.com/          javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
https://tls13.akamai.io/                 java.net.ConnectException: Failed to connect to tls13.akamai.io/2a02:26f0:5c:19f:0:0:0:5f37:443
https://swifttls.org/                    TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.0 200 222b
https://www.googleapis.com/robots.txt    java.net.ConnectException: Failed to connect to www.googleapis.com/2a00:1450:4009:801:0:0:0:200a:443
https://graph.facebook.com/robots.txt    TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 25b
https://api.twitter.com/robots.txt       javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
https://connect.squareup.com/robots.txt  javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure

TLS1.3 then fallback
https://enabled.tls13.com                TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 h2 200 140b
https://www.howsmyssl.com/a/check        TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 468b
https://tls13.cloudflare.com             TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 h2 521 4546b
https://www.allizom.org/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 h2 200 125b
https://tls13.crypto.mozilla.org/        javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://tls.ctf.network/robots.txt       TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 404 169b
https://rustls.jbp.io/                   TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 200 416b
https://h2o.examp1e.net                  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 h2 200 6209b
https://mew.org/                         TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 137b
https://tls13.baishancloud.com/          TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 2737b
https://tls13.akamai.io/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 4324b
https://swifttls.org/                    TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.0 200 222b
https://www.googleapis.com/robots.txt    javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://graph.facebook.com/robots.txt    TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 25b
https://api.twitter.com/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 h2 200 155b
https://connect.squareup.com/robots.txt  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 404 0b

TLS1.3+TLS1.2
https://enabled.tls13.com                TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 h2 200 140b
https://www.howsmyssl.com/a/check        TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 518b
https://tls13.cloudflare.com             TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 h2 521 4546b
https://www.allizom.org/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 h2 200 125b
https://tls13.crypto.mozilla.org/        javax.net.ssl.SSLHandshakeException: Received fatal alert: protocol_version
https://tls.ctf.network/robots.txt       TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 404 169b
https://rustls.jbp.io/                   TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.1 200 416b
https://h2o.examp1e.net                  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 h2 200 6209b
https://mew.org/                         TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 137b
https://tls13.baishancloud.com/          javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
https://tls13.akamai.io/                 java.net.ConnectException: Failed to connect to tls13.akamai.io/2a02:26f0:5c:19f:0:0:0:5f37:443
https://swifttls.org/                    TLS_1_3 TLS_AES_128_GCM_SHA256 http/1.0 200 222b
https://www.googleapis.com/robots.txt    javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
https://graph.facebook.com/robots.txt    TLS_1_3 TLS_AES_128_GCM_SHA256 h2 200 25b
https://api.twitter.com/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 h2 200 155b
https://connect.squareup.com/robots.txt  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 404 0b
```

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/ConnectionSpec.java]: Results with JDK 8

```
Running tests using Platform 25.162-b12
TLS1.3 only
https://enabled.tls13.com                java.net.ConnectException: Failed to connect to enabled.tls13.com/2400:cb00:2048:1:0:0:6810:7e22:443
https://www.howsmyssl.com/a/check        java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://tls13.cloudflare.com             java.net.ConnectException: Failed to connect to tls13.cloudflare.com/2400:cb00:2048:1:0:0:6813:c597:443
https://www.allizom.org/robots.txt       java.net.ConnectException: Failed to connect to www.allizom.org/2400:cb00:2048:1:0:0:6810:7e21:443
https://tls13.crypto.mozilla.org/        java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://tls.ctf.network/robots.txt       java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://rustls.jbp.io/                   java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://h2o.examp1e.net                  java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://mew.org/                         java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://tls13.baishancloud.com/          java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://tls13.akamai.io/                 java.net.ConnectException: Failed to connect to tls13.akamai.io/2a02:26f0:5c:1a2:0:0:0:5f37:443
https://swifttls.org/                    java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://www.googleapis.com/robots.txt    java.net.ConnectException: Failed to connect to www.googleapis.com/2a00:1450:4009:801:0:0:0:200a:443
https://graph.facebook.com/robots.txt    java.net.ConnectException: Failed to connect to graph.facebook.com/2a03:2880:f029:11:face:b00c:0:2:443
https://api.twitter.com/robots.txt       java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]
https://connect.squareup.com/robots.txt  java.net.UnknownServiceException: Unable to find acceptable protocols. isFallback=false, modes=[ConnectionSpec(cipherSuites=[TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_CCM_SHA256, TLS_AES_256_CCM_8_SHA256], tlsVersions=[TLS_1_3], supportsTlsExtensions=true)], supported protocols=[TLSv1, TLSv1.1, TLSv1.2]

TLS1.3 then fallback
https://enabled.tls13.com                TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 200 140b
https://www.howsmyssl.com/a/check        TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 468b
https://tls13.cloudflare.com             TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 521 4546b
https://www.allizom.org/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 125b
https://tls13.crypto.mozilla.org/        javax.net.ssl.SSLException: Received fatal alert: protocol_version
https://tls.ctf.network/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 404 169b
https://rustls.jbp.io/                   TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 416b
https://h2o.examp1e.net                  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 6209b
https://mew.org/                         TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 137b
https://tls13.baishancloud.com/          TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 2737b
https://tls13.akamai.io/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 4312b
https://swifttls.org/                    TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.0 200 237b
https://www.googleapis.com/robots.txt    TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 200 25b
https://graph.facebook.com/robots.txt    TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 200 25b
https://api.twitter.com/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 155b
https://connect.squareup.com/robots.txt  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 404 0b

TLS1.3+TLS1.2
https://enabled.tls13.com                TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 200 140b
https://www.howsmyssl.com/a/check        TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 468b
https://tls13.cloudflare.com             TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 521 4546b
https://www.allizom.org/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 125b
https://tls13.crypto.mozilla.org/        javax.net.ssl.SSLException: Received fatal alert: protocol_version
https://tls.ctf.network/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 404 169b
https://rustls.jbp.io/                   TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 416b
https://h2o.examp1e.net                  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 6209b
https://mew.org/                         TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 137b
https://tls13.baishancloud.com/          TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 2737b
https://tls13.akamai.io/                 TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.1 200 4312b
https://swifttls.org/                    TLS_1_2 TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 http/1.0 200 237b
https://www.googleapis.com/robots.txt    TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 200 25b
https://graph.facebook.com/robots.txt    TLS_1_2 TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 http/1.1 200 25b
https://api.twitter.com/robots.txt       TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 200 155b
https://connect.squareup.com/robots.txt  TLS_1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 http/1.1 404 0b
```

**Comment by yschimke** [comment]
Fixed in JDK11 27 - https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8210002

**Comment by yschimke** [comment]
@swankjesse Anything needed here?

File: okhttp-tests/src/main/java/okhttp3/TestTls13Request.java [hunk 2]
```diff
@@ -9,13 +9,36 @@
...
+      CipherSuite.TLS_AES_256_GCM_SHA384,
+      CipherSuite.TLS_CHACHA20_POLY1305_SHA256,
+      CipherSuite.TLS_AES_128_CCM_SHA256,
+      CipherSuite.TLS_AES_256_CCM_8_SHA256
+  };
+
+  /**
+   * A TLS 1.3 only Connection Spec. This will be eventually be exposed
+   * as part of MODERN_TLS or folded into the default OkHttp client once published and
+   * available in JDK11 or Conscrypt.
```
Comment (swankjesse) [inline okhttp-tests/src/main/java/okhttp3/TestTls13Request.java]: 👍

File: okhttp/src/main/java/okhttp3/CipherSuite.java [hunk 3]
```diff
@@ -395,8 +395,7 @@
   // public static final CipherSuite TLS_DHE_PSK_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_DHE_PSK_WITH_CHACHA20_POLY1305_SHA256", 0xccad);
   // public static final CipherSuite TLS_RSA_PSK_WITH_CHACHA20_POLY1305_SHA256 = of("TLS_RSA_PSK_WITH_CHACHA20_POLY1305_SHA256", 0xccae);
 
-  // TLS 1.3 https://tools.ietf.org/html/draft-ietf-tls-tls13-28
-  // Documentation only ids will be updated once assigned by IANA.
+  // TLS 1.3 https://tools.ietf.org/html/rfc8446
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/CipherSuite.java]: Until now I hadn’t even realized that these don’t overlap with the other ones. That’s actually going to be very handy!

**Comment by swankjesse** [comment]
This is fantastic. Thanks and thanks for your patience.

File: okhttp-tests/src/main/java/okhttp3/TestTls13Request.java (hunk 2 above)
Comment (yschimke) [reply okhttp-tests/src/main/java/okhttp3/TestTls13Request.java]: That's now out of date :(

File: okhttp/src/main/java/okhttp3/CipherSuite.java (hunk 3 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/CipherSuite.java]: Awkwardly, Conscrypt will ignore and only use the first three for now, no matter what you ask for.

File: okhttp/src/main/java/okhttp3/ConnectionSpec.java [hunk 4]
```diff
@@ -50,19 +57,17 @@
       CipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
   };
 
-  // TLS 1.3
-  private static final CipherSuite[] TLS13_CIPHER_SUITES = new CipherSuite[] {
+  // This is nearly equal to the cipher suites supported in Chrome 51, current as of 2016-05-25.
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/ConnectionSpec.java]: I should update this also.

**Pull request merged**